File: apps/lesson-plan-generator/internal/pipeline/scenes.go
package pipeline

import "lesson-plan-generator/internal/domain"

type sceneEntry struct {
	script   domain.SceneScript
	duration float64
}

// step 5 – map scene scripts to (script, duration) pairs.
// Duration = sum of assigned audio segment durations, or data["duration"] for narration-free scenes.
func generateScenes(scripts []domain.SceneScript, audio []domain.AudioSegment) []sceneEntry {
	durMap := make(map[string]float64, len(audio))
	for _, s := range audio {
		durMap[s.SegmentID] = s.DurationSec
	}

	entries := make([]sceneEntry, 0, len(scripts))
	for _, script := range scripts {
		var duration float64
		if len(script.NarrationSegmentIDs) > 0 {
			for _, id := range script.NarrationSegmentIDs {
				duration += durMap[id]
			}
		} else {
			duration = getFloat(script.Data, "duration", 3.0)
		}
		entries = append(entries, sceneEntry{script: script, duration: duration})
	}
	return entries
}

File: apps/lesson-plan-generator/app/steps/scenes.py
"""Step 5 – Map scene scripts to (script, duration) pairs.

Duration for scenes with narration = sum of their assigned audio segment durations.
Duration for scenes without narration (e.g. title_scene) = data["duration"] field.
"""

from app.models import AudioSegment, SceneScript


def generate_scenes(
    scene_scripts: list[SceneScript],
    audio_segments: list[AudioSegment],
) -> list[tuple[SceneScript, float]]:
    seg_dur: dict[str, float] = {s.segment_id: s.duration_sec for s in audio_segments}

    result: list[tuple[SceneScript, float]] = []
    for script in scene_scripts:
        if script.narration_segment_ids:
            duration = sum(seg_dur[sid] for sid in script.narration_segment_ids if sid in seg_dur)
        else:
            duration = float(script.data.get("duration", 3.0))
        result.append((script, duration))

    return result
